In [976]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [977]:
data=pd.read_csv('/content/HINDALCO.csv')

In [978]:
data.head()

Date    Symbol Series  Prev Close     Open     High     Low    Last  \
0  2000-01-03  HINDALC0     EQ      813.45   851.00   878.50   851.0   869.0   
1  2000-01-04  HINDALC0     EQ      867.45   879.00   930.00   870.0   900.0   
2  2000-01-05  HINDALC0     EQ      900.35   835.00   972.40   835.0   972.4   
3  2000-01-06  HINDALC0     EQ      972.35  1050.15  1050.15  1015.1  1045.0   
4  2000-01-07  HINDALC0     EQ     1048.45  1096.00  1120.00  1015.0  1086.0   

     Close     VWAP  Volume      Turnover  Trades  Deliverable Volume  \
0   867.45   869.86   33799  2.940055e+12     NaN                 NaN   
1   900.35   903.51   70944  6.409848e+12     NaN                 NaN   
2   972.35   949.32  153030  1.452744e+13     NaN                 NaN   
3  1048.45  1047.08  129708  1.358144e+13     NaN                 NaN   
4  1094.95  1085.24  138829  1.506623e+13     NaN                 NaN   

   %Deliverble  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN



*   Volume,Turnover,Trades are the metrics indicating trading activity
*   Prev Close, Open, High, Low, Last, Close: Stock price data.


*   Deliverable Volume, %Deliverble: Indicators of stock deliverability.





In [979]:
data.shape

(5306, 15)

 Analyze the indicators in the dataset that best explain the volatility and
 unpredictable nature of the stocks in the last decade.



*  Calculating volatility metrics:High-low % range,Daily Change %




In [980]:
data['Date']=pd.to_datetime(data['Date'])

In [981]:
data.head()

Date    Symbol Series  Prev Close     Open     High     Low    Last  \
0 2000-01-03  HINDALC0     EQ      813.45   851.00   878.50   851.0   869.0   
1 2000-01-04  HINDALC0     EQ      867.45   879.00   930.00   870.0   900.0   
2 2000-01-05  HINDALC0     EQ      900.35   835.00   972.40   835.0   972.4   
3 2000-01-06  HINDALC0     EQ      972.35  1050.15  1050.15  1015.1  1045.0   
4 2000-01-07  HINDALC0     EQ     1048.45  1096.00  1120.00  1015.0  1086.0   

     Close     VWAP  Volume      Turnover  Trades  Deliverable Volume  \
0   867.45   869.86   33799  2.940055e+12     NaN                 NaN   
1   900.35   903.51   70944  6.409848e+12     NaN                 NaN   
2   972.35   949.32  153030  1.452744e+13     NaN                 NaN   
3  1048.45  1047.08  129708  1.358144e+13     NaN                 NaN   
4  1094.95  1085.24  138829  1.506623e+13     NaN                 NaN   

   %Deliverble  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN

In [982]:
data['High-Low Range %'] = ((data['High'] - data['Low']) / data['Low']) * 100


In [983]:
data['Daily Change %'] = ((data['Close'] - data['Prev Close']) / data['Prev Close']) * 100


* Daily Change %: Indicates the percentage change in stock price from the previous close to the current close.
*  High-Low Range %:Represents the daily range as a percentage of the closing price.
The average daily range is approximately 3.54%, with a standard deviation of 2.55%.
Maximum observed daily range is 93.56%, suggesting highly volatile days.





In [984]:
df=data.loc[:,'Prev Close':]

In [985]:
df.head()

Prev Close     Open     High     Low    Last    Close     VWAP  Volume  \
0      813.45   851.00   878.50   851.0   869.0   867.45   869.86   33799   
1      867.45   879.00   930.00   870.0   900.0   900.35   903.51   70944   
2      900.35   835.00   972.40   835.0   972.4   972.35   949.32  153030   
3      972.35  1050.15  1050.15  1015.1  1045.0  1048.45  1047.08  129708   
4     1048.45  1096.00  1120.00  1015.0  1086.0  1094.95  1085.24  138829   

       Turnover  Trades  Deliverable Volume  %Deliverble  High-Low Range %  \
0  2.940055e+12     NaN                 NaN          NaN          3.231492   
1  6.409848e+12     NaN                 NaN          NaN          6.896552   
2  1.452744e+13     NaN                 NaN          NaN         16.455090   
3  1.358144e+13     NaN                 NaN          NaN          3.452862   
4  1.506623e+13     NaN                 NaN          NaN         10.344828   

   Daily Change %  
0        6.638392  
1        3.792726  
2        7.996890  
3        7.826400  
4        4.435119

In [986]:
new_df=data[(data['Date']>'2014-12-01')& (data['Date']<'2024-12-02')]#dataset of last decade

In [987]:
new_df.head()

Date    Symbol Series  Prev Close    Open    High    Low   Last  \
3719 2014-12-02  HINDALCO     EQ      167.40  167.00  172.25  166.0  171.8   
3720 2014-12-03  HINDALCO     EQ      171.05  171.80  173.00  168.0  168.6   
3721 2014-12-04  HINDALCO     EQ      169.05  170.05  171.30  165.3  165.7   
3722 2014-12-05  HINDALCO     EQ      165.90  167.45  167.80  162.9  163.3   
3723 2014-12-08  HINDALCO     EQ      163.20  162.75  164.85  157.8  158.0   

       Close    VWAP   Volume      Turnover   Trades  Deliverable Volume  \
3719  171.05  168.98  9360449  1.581767e+14  68490.0           4228884.0   
3720  169.05  170.45  6342942  1.081159e+14  50205.0           2539858.0   
3721  165.90  167.24  7635372  1.276910e+14  40458.0           3329526.0   
3722  163.20  164.90  6802949  1.121817e+14  39320.0           3999378.0   
3723  159.15  160.13  9025091  1.445165e+14  71746.0           5618517.0   

      %Deliverble  High-Low Range %  Daily Change %  
3719       0.4518          3.765060        2.180406  
3720       0.4004          2.976190       -1.169249  
3721       0.4361          3.629764       -1.863354  
3722       0.5879          3.007980       -1.627486  
3723       0.6225          4.467681       -2.481618

In [988]:
df=new_df.loc[:,'Prev Close':]

In [989]:
df.corr()

Prev Close      Open      High       Low      Last  \
Prev Close            1.000000  0.999455  0.998359  0.998293  0.997121   
Open                  0.999455  1.000000  0.998919  0.998962  0.997728   
High                  0.998359  0.998919  1.000000  0.998926  0.999155   
Low                   0.998293  0.998962  0.998926  1.000000  0.999020   
Last                  0.997121  0.997728  0.999155  0.999020  1.000000   
Close                 0.997127  0.997754  0.999185  0.999088  0.999959   
VWAP                  0.998168  0.998789  0.999686  0.999593  0.999545   
Volume                0.089928  0.093759  0.116254  0.091264  0.108524   
Turnover              0.534184  0.537803  0.558111  0.537108  0.551816   
Trades                0.411471  0.413541  0.432716  0.408802  0.423608   
Deliverable Volume    0.061427  0.063220  0.075339  0.057944  0.068032   
%Deliverble          -0.052452 -0.053731 -0.062740 -0.056098 -0.063121   
High-Low Range %     -0.177390 -0.180336 -0.159172 -0.201361 -0.177991   
Daily Change %       -0.017160 -0.002557  0.027526  0.027664  0.052477   

                       Close      VWAP    Volume  Turnover    Trades  \
Prev Close          0.997127  0.998168  0.089928  0.534184  0.411471   
Open                0.997754  0.998789  0.093759  0.537803  0.413541   
High                0.999185  0.999686  0.116254  0.558111  0.432716   
Low                 0.999088  0.999593  0.091264  0.537108  0.408802   
Last                0.999959  0.999545  0.108524  0.551816  0.423608   
Close               1.000000  0.999603  0.108538  0.552007  0.423669   
VWAP                0.999603  1.000000  0.106616  0.550549  0.422989   
Volume              0.108538  0.106616  1.000000  0.838115  0.841525   
Turnover            0.552007  0.550549  0.838115  1.000000  0.890434   
Trades              0.423669  0.422989  0.841525  0.890434  1.000000   
Deliverable Volume  0.067748  0.067301  0.776018  0.630169  0.634264   
%Deliverble        -0.063556 -0.061892 -0.127028 -0.121282 -0.108981   
High-Low Range %   -0.178809 -0.178930  0.476890  0.268142  0.386971   
Daily Change %      0.053022  0.035167  0.243524  0.202696  0.148044   

                    Deliverable Volume  %Deliverble  High-Low Range %  \
Prev Close                    0.061427    -0.052452         -0.177390   
Open                          0.063220    -0.053731         -0.180336   
High                          0.075339    -0.062740         -0.159172   
Low                           0.057944    -0.056098         -0.201361   
Last                          0.068032    -0.063121         -0.177991   
Close                         0.067748    -0.063556         -0.178809   
VWAP                          0.067301    -0.061892         -0.178930   
Volume                        0.776018    -0.127028          0.476890   
Turnover                      0.630169    -0.121282          0.268142   
Trades                        0.634264    -0.108981          0.386971   
Deliverable Volume            1.000000     0.433880          0.350351   
%Deliverble                   0.433880     1.000000         -0.114414   
High-Low Range %              0.350351    -0.114414          1.000000   
Daily Change %                0.087634    -0.156867         -0.024093   

                    Daily Change %  
Prev Close               -0.017160  
Open                     -0.002557  
High                      0.027526  
Low                       0.027664  
Last                      0.052477  
Close                     0.053022  
VWAP                      0.035167  
Volume                    0.243524  
Turnover                  0.202696  
Trades                    0.148044  
Deliverable Volume        0.087634  
%Deliverble              -0.156867  
High-Low Range %         -0.024093  
Daily Change %            1.000000

Correlation Analysis
High-Low Range %--



* Trades:Positive Correlation(0.386) between High-low range(%) and Trades
*  Volume:Positive Correlation(0.4768).Increase in Volume lead to increase in Trades.  

*  Turnover:Positive Correlation(0.268).Increase in Turnover lead to increase in High-Low Range
*  Deliverable%:Negative Correlation(-0.114) between Deliverable Volume and High ow Range%








Daily Change % (Price Change Volatility):
* Turnover:Positive correlation(0.202) suggest a slight relationship between Turnover and Daily Change%.
* Trades: Positive correlation(0.148) suggest a slight relationship between Turnover and Daily Change%.

* Volume:Positive Correlation(0.243) suggest a positive relationship
*  Deliverable Volume:Positive Correlation(0.087)suggest a positive relationship between Deliverable Volume and Daily Change%.
   

Insights

1.  Trades and Volume are critical indicators of intraday volatility.
2.  Turnover slightly affects price volatility but has a stronger impact on intraday volatility.
3.%Deliverable inversely correlates with both volatility measures, potentially indicating that stocks with more speculative trading (lower deliverability) are more volatile

In [990]:
data.isnull().sum()

Date                     0
Symbol                   0
Series                   0
Prev Close               0
Open                     0
High                     0
Low                      0
Last                     0
Close                    0
VWAP                     0
Volume                   0
Turnover                 0
Trades                2850
Deliverable Volume     525
%Deliverble            525
High-Low Range %         0
Daily Change %           0
dtype: int64

In [991]:
for col in data.columns:
  if data[col].dtype=='int64' or data[col].dtype=='float64':
    data[col]=data[col].fillna(data[col].mean())

In [992]:
data.isna().sum()

Date                  0
Symbol                0
Series                0
Prev Close            0
Open                  0
High                  0
Low                   0
Last                  0
Close                 0
VWAP                  0
Volume                0
Turnover              0
Trades                0
Deliverable Volume    0
%Deliverble           0
High-Low Range %      0
Daily Change %        0
dtype: int64

#Choose promising stock based on your analysis for your portfolio

In [993]:
data

Date    Symbol Series  Prev Close     Open     High     Low  \
0    2000-01-03  HINDALC0     EQ      813.45   851.00   878.50   851.0   
1    2000-01-04  HINDALC0     EQ      867.45   879.00   930.00   870.0   
2    2000-01-05  HINDALC0     EQ      900.35   835.00   972.40   835.0   
3    2000-01-06  HINDALC0     EQ      972.35  1050.15  1050.15  1015.1   
4    2000-01-07  HINDALC0     EQ     1048.45  1096.00  1120.00  1015.0   
...         ...       ...    ...         ...      ...      ...     ...   
5301 2021-04-26  HINDALCO     EQ      349.50   353.00   358.25   345.5   
5302 2021-04-27  HINDALCO     EQ      348.35   350.10   370.00   350.1   
5303 2021-04-28  HINDALCO     EQ      366.25   368.85   368.90   358.0   
5304 2021-04-29  HINDALCO     EQ      362.60   367.40   377.30   364.6   
5305 2021-04-30  HINDALCO     EQ      372.15   368.90   377.00   361.3   

         Last    Close     VWAP    Volume      Turnover         Trades  \
0      869.00   867.45   869.86     33799  2.940055e+12   60639.510586   
1      900.00   900.35   903.51     70944  6.409848e+12   60639.510586   
2      972.40   972.35   949.32    153030  1.452744e+13   60639.510586   
3     1045.00  1048.45  1047.08    129708  1.358144e+13   60639.510586   
4     1086.00  1094.95  1085.24    138829  1.506623e+13   60639.510586   
...       ...      ...      ...       ...           ...            ...   
5301   348.60   348.35   351.91  16472269  5.796822e+14  113431.000000   
5302   366.00   366.25   365.09  41263319  1.506467e+15  307599.000000   
5303   362.45   362.60   361.61  15197329  5.495571e+14  126873.000000   
5304   370.50   372.15   370.70  34391622  1.274898e+15  237157.000000   
5305   363.65   364.40   368.96  19935239  7.355374e+14  149528.000000   

      Deliverable Volume  %Deliverble  High-Low Range %  Daily Change %  
0           2.551709e+06     0.434026          3.231492        6.638392  
1           2.551709e+06     0.434026          6.896552        3.792726  
2           2.551709e+06     0.434026         16.455090        7.996890  
3           2.551709e+06     0.434026          3.452862        7.826400  
4           2.551709e+06     0.434026         10.344828        4.435119  
...                  ...          ...               ...             ...  
5301        7.186409e+06     0.436300          3.690304       -0.329041  
5302        8.981182e+06     0.217700          5.684090        5.138510  
5303        3.077711e+06     0.202500          3.044693       -0.996587  
5304        6.771282e+06     0.196900          3.483269        2.633756  
5305        4.268054e+06     0.214100          4.345419       -2.082494  

[5306 rows x 17 columns]

In [994]:
data.columns

Index(['Date', 'Symbol', 'Series', 'Prev Close', 'Open', 'High', 'Low', 'Last',
       'Close', 'VWAP', 'Volume', 'Turnover', 'Trades', 'Deliverable Volume',
       '%Deliverble', 'High-Low Range %', 'Daily Change %'],
      dtype='object')

In [995]:
# Calculate summary statistics by stock
stock_analysis = data.groupby('Symbol').agg({
    'Daily Change %': ['mean', 'std', lambda x: (x > 0).sum() / len(x)],  # Mean, Std Dev, and Positive Return Frequency
    'High-Low Range %': ['mean'],                                           # Mean volatility
    'Volume': ['mean'],                                                   # Mean trading volume
    '%Deliverble': ['mean'],                                              # Mean deliverable percentage
    'VWAP': ['mean']                                                      # Average VWAP
})

# Flatten column names
stock_analysis.columns = [
    'Mean Return (%)', 'Return Std Dev (%)', 'Positive Return Frequency',
    'Mean Volatility (%)', 'Mean Volume', 'Mean Deliverable (%)', 'Mean VWAP'
]

# Define thresholds or scores for the criteria
stock_analysis['Consistency Score'] = stock_analysis['Positive Return Frequency'] * 100  # Positive return frequency as a percentage
stock_analysis['Moderate Volatility Score'] = -abs(stock_analysis['Mean Volatility (%)'] - stock_analysis['Mean Volatility (%)'].median())
stock_analysis['High Volume Score'] = stock_analysis['Mean Volume'] / stock_analysis['Mean Volume'].max() * 100  # Volume as a percentage of the max volume
stock_analysis['Trend Score'] = stock_analysis['Mean Deliverable (%)'] + stock_analysis['Mean VWAP']  # Simplified trend measure

# Combine scores for overall ranking
stock_analysis['Overall Score'] = (
    stock_analysis['Consistency Score'] +
    stock_analysis['Moderate Volatility Score'] +
    stock_analysis['High Volume Score'] +
    stock_analysis['Trend Score']
)

# Rank stocks based on the overall score
stock_analysis['Rank'] = stock_analysis['Overall Score'].rank(ascending=False)

# Select the top stocks based on overall rank
top_consistent_stocks = stock_analysis.sort_values('Rank').head(5).reset_index()

# Display the top stocks
print(top_consistent_stocks[['Symbol', 'Mean Return (%)', 'Positive Return Frequency',
                             'Mean Volatility (%)', 'Mean Volume',
                             'Mean Deliverable (%)', 'Mean VWAP', 'Overall Score']])


     Symbol  Mean Return (%)  Positive Return Frequency  Mean Volatility (%)  \
0  HINDALC0         0.007016                   0.504950             3.846889   
1  HINDALCO         0.070118                   0.500148             4.162188   

    Mean Volume  Mean Deliverable (%)   Mean VWAP  Overall Score  
0  1.092017e+06              0.552352  702.822209     765.475780  
1  9.282843e+06              0.366985  159.308854     309.532952  


In [996]:
data.head()

Date    Symbol Series  Prev Close     Open     High     Low    Last  \
0 2000-01-03  HINDALC0     EQ      813.45   851.00   878.50   851.0   869.0   
1 2000-01-04  HINDALC0     EQ      867.45   879.00   930.00   870.0   900.0   
2 2000-01-05  HINDALC0     EQ      900.35   835.00   972.40   835.0   972.4   
3 2000-01-06  HINDALC0     EQ      972.35  1050.15  1050.15  1015.1  1045.0   
4 2000-01-07  HINDALC0     EQ     1048.45  1096.00  1120.00  1015.0  1086.0   

     Close     VWAP  Volume      Turnover        Trades  Deliverable Volume  \
0   867.45   869.86   33799  2.940055e+12  60639.510586        2.551709e+06   
1   900.35   903.51   70944  6.409848e+12  60639.510586        2.551709e+06   
2   972.35   949.32  153030  1.452744e+13  60639.510586        2.551709e+06   
3  1048.45  1047.08  129708  1.358144e+13  60639.510586        2.551709e+06   
4  1094.95  1085.24  138829  1.506623e+13  60639.510586        2.551709e+06   

   %Deliverble  High-Low Range %  Daily Change %  
0     0.434026          3.231492        6.638392  
1     0.434026          6.896552        3.792726  
2     0.434026         16.455090        7.996890  
3     0.434026          3.452862        7.826400  
4     0.434026         10.344828        4.435119

In [997]:
data

Date    Symbol Series  Prev Close     Open     High     Low  \
0    2000-01-03  HINDALC0     EQ      813.45   851.00   878.50   851.0   
1    2000-01-04  HINDALC0     EQ      867.45   879.00   930.00   870.0   
2    2000-01-05  HINDALC0     EQ      900.35   835.00   972.40   835.0   
3    2000-01-06  HINDALC0     EQ      972.35  1050.15  1050.15  1015.1   
4    2000-01-07  HINDALC0     EQ     1048.45  1096.00  1120.00  1015.0   
...         ...       ...    ...         ...      ...      ...     ...   
5301 2021-04-26  HINDALCO     EQ      349.50   353.00   358.25   345.5   
5302 2021-04-27  HINDALCO     EQ      348.35   350.10   370.00   350.1   
5303 2021-04-28  HINDALCO     EQ      366.25   368.85   368.90   358.0   
5304 2021-04-29  HINDALCO     EQ      362.60   367.40   377.30   364.6   
5305 2021-04-30  HINDALCO     EQ      372.15   368.90   377.00   361.3   

         Last    Close     VWAP    Volume      Turnover         Trades  \
0      869.00   867.45   869.86     33799  2.940055e+12   60639.510586   
1      900.00   900.35   903.51     70944  6.409848e+12   60639.510586   
2      972.40   972.35   949.32    153030  1.452744e+13   60639.510586   
3     1045.00  1048.45  1047.08    129708  1.358144e+13   60639.510586   
4     1086.00  1094.95  1085.24    138829  1.506623e+13   60639.510586   
...       ...      ...      ...       ...           ...            ...   
5301   348.60   348.35   351.91  16472269  5.796822e+14  113431.000000   
5302   366.00   366.25   365.09  41263319  1.506467e+15  307599.000000   
5303   362.45   362.60   361.61  15197329  5.495571e+14  126873.000000   
5304   370.50   372.15   370.70  34391622  1.274898e+15  237157.000000   
5305   363.65   364.40   368.96  19935239  7.355374e+14  149528.000000   

      Deliverable Volume  %Deliverble  High-Low Range %  Daily Change %  
0           2.551709e+06     0.434026          3.231492        6.638392  
1           2.551709e+06     0.434026          6.896552        3.792726  
2           2.551709e+06     0.434026         16.455090        7.996890  
3           2.551709e+06     0.434026          3.452862        7.826400  
4           2.551709e+06     0.434026         10.344828        4.435119  
...                  ...          ...               ...             ...  
5301        7.186409e+06     0.436300          3.690304       -0.329041  
5302        8.981182e+06     0.217700          5.684090        5.138510  
5303        3.077711e+06     0.202500          3.044693       -0.996587  
5304        6.771282e+06     0.196900          3.483269        2.633756  
5305        4.268054e+06     0.214100          4.345419       -2.082494  

[5306 rows x 17 columns]

 Feature engineer and build machine learning solutions for the business
 Requirements

In [998]:
data['Cumulative_Return'] = data['Close'] / data['Close'].iloc[0] - 1  #cumilative returns over a period


In [999]:
data['Volume_Change'] = data['Volume'].pct_change()


In [1000]:
#Moving averages
data['SMA_50'] = data['Close'].rolling(window=50).mean()
data['SMA_200'] = data['Close'].rolling(window=200).mean()


In [1001]:
data['Volatility'] = data['Daily Change %'].rolling(window=20).std()


In [1002]:
data.head()

Date    Symbol Series  Prev Close     Open     High     Low    Last  \
0 2000-01-03  HINDALC0     EQ      813.45   851.00   878.50   851.0   869.0   
1 2000-01-04  HINDALC0     EQ      867.45   879.00   930.00   870.0   900.0   
2 2000-01-05  HINDALC0     EQ      900.35   835.00   972.40   835.0   972.4   
3 2000-01-06  HINDALC0     EQ      972.35  1050.15  1050.15  1015.1  1045.0   
4 2000-01-07  HINDALC0     EQ     1048.45  1096.00  1120.00  1015.0  1086.0   

     Close     VWAP  ...        Trades  Deliverable Volume  %Deliverble  \
0   867.45   869.86  ...  60639.510586        2.551709e+06     0.434026   
1   900.35   903.51  ...  60639.510586        2.551709e+06     0.434026   
2   972.35   949.32  ...  60639.510586        2.551709e+06     0.434026   
3  1048.45  1047.08  ...  60639.510586        2.551709e+06     0.434026   
4  1094.95  1085.24  ...  60639.510586        2.551709e+06     0.434026   

   High-Low Range %  Daily Change %  Cumulative_Return  Volume_Change  SMA_50  \
0          3.231492        6.638392           0.000000            NaN     NaN   
1          6.896552        3.792726           0.037927       1.098997     NaN   
2         16.455090        7.996890           0.120929       1.157053     NaN   
3          3.452862        7.826400           0.208658      -0.152401     NaN   
4         10.344828        4.435119           0.262263       0.070319     NaN   

   SMA_200  Volatility  
0      NaN         NaN  
1      NaN         NaN  
2      NaN         NaN  
3      NaN         NaN  
4      NaN         NaN  

[5 rows x 22 columns]

In [1003]:
data['Target_Close'] = data['Close'].shift(-1)


In [1004]:
data.head()

Date    Symbol Series  Prev Close     Open     High     Low    Last  \
0 2000-01-03  HINDALC0     EQ      813.45   851.00   878.50   851.0   869.0   
1 2000-01-04  HINDALC0     EQ      867.45   879.00   930.00   870.0   900.0   
2 2000-01-05  HINDALC0     EQ      900.35   835.00   972.40   835.0   972.4   
3 2000-01-06  HINDALC0     EQ      972.35  1050.15  1050.15  1015.1  1045.0   
4 2000-01-07  HINDALC0     EQ     1048.45  1096.00  1120.00  1015.0  1086.0   

     Close     VWAP  ...  Deliverable Volume  %Deliverble  High-Low Range %  \
0   867.45   869.86  ...        2.551709e+06     0.434026          3.231492   
1   900.35   903.51  ...        2.551709e+06     0.434026          6.896552   
2   972.35   949.32  ...        2.551709e+06     0.434026         16.455090   
3  1048.45  1047.08  ...        2.551709e+06     0.434026          3.452862   
4  1094.95  1085.24  ...        2.551709e+06     0.434026         10.344828   

   Daily Change %  Cumulative_Return  Volume_Change  SMA_50  SMA_200  \
0        6.638392           0.000000            NaN     NaN      NaN   
1        3.792726           0.037927       1.098997     NaN      NaN   
2        7.996890           0.120929       1.157053     NaN      NaN   
3        7.826400           0.208658      -0.152401     NaN      NaN   
4        4.435119           0.262263       0.070319     NaN      NaN   

   Volatility  Target_Close  
0         NaN        900.35  
1         NaN        972.35  
2         NaN       1048.45  
3         NaN       1094.95  
4         NaN       1022.90  

[5 rows x 23 columns]

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score
import matplotlib.pyplot as plt


# Ensure the Date column is in datetime format
data['Date'] = pd.to_datetime(data['Date'])


# Feature Engineering
data['Daily_Return'] = data['Close'].pct_change()
data['SMA_20'] = data['Close'].rolling(window=20).mean()
data['SMA_50'] = data['Close'].rolling(window=50).mean()
data['Volatility'] = data['Close'].rolling(window=20).std()

# Drop rows with NaN values generated from rolling calculations
data.dropna(inplace=True)
# Drop the last row with NaN target
data = data.dropna()

# Define Features (X) and Target (y)
features = ['Open', 'High', 'Low', 'Close', 'Volume', 'Daily_Return', 'SMA_20', 'SMA_50', 'Volatility']
X = data[features]
y = data['Target_Close']

# Train-Test Split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train Random Forest Regressor
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

# Make Predictions
y_pred = rf_model.predict(X_test)

# Evaluate the Model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f"Mean Squared Error: {mse}")
print(f"R2 Score: {r2}")

# Plot Actual vs Predicted Prices
plt.figure(figsize=(10, 6))
plt.plot(y_test.values, label="Actual Prices", color="blue")
plt.plot(y_pred, label="Predicted Prices", color="orange")
plt.title("Actual vs Predicted Prices")
plt.xlabel("Time")
plt.ylabel("Stock Price")
plt.legend()
plt.show()


In [ ]:
data.head()

In [ ]:
features=['Date','Close']

In [ ]:
my_data=data[features]

In [ ]:
my_data['Date']=pd.to_datetime(my_data['Date'])

In [ ]:
my_data

In [ ]:
my_data.index=my_data['Date']

In [ ]:
del my_data['Date']

In [ ]:
my_data.plot()

In [ ]:
std_log=my_data.rolling(window=4).std()

In [ ]:
mean_log=my_data.rolling(window=4).mean()

In [ ]:
plt.plot(my_data,color='blue',label='Original Data')
plt.plot(mean_log,color='black',label='Rolling Mean')
plt.plot(std_log,color='red',label='Rolling Standard Deviation')
plt.legend(loc='best')

In [ ]:
from statsmodels.tsa.stattools  import adfuller

In [ ]:
#Augmented DickyFuller Test
result=adfuller(my_data['Close'])
keys=['test_statistics','p_value',"num_lags","num_observations","critical_values","AIC"]

In [ ]:
for key,value in zip(keys,result):
  print(f"{key}:{value}")


In [ ]:
import numpy as np

In [ ]:
first_log=np.log(my_data)

In [ ]:
first_log.dropna()

In [ ]:
first_log.plot()

In [ ]:
new_data=first_log-mean_log

In [ ]:
mean_log=new_data.rolling(window=12).mean()
std_log=new_data.rolling(window=12).std()

plt.plot(new_data,color='blue',label='Original data')
plt.plot(mean_log,color='red',label='Mean data')
plt.plot(std_log,color='black',label='Standard Deviation')

plt.legend(loc='best')
plt.title("Rolling mean and Standard deviation for log_scale data")

In [ ]:
new_data

In [ ]:
from statsmodels.tsa.stattools import acf,pacf
from statsmodels.graphics.tsaplots import plot_acf,plot_pacf

In [ ]:
acf_values=acf(new_data)
plot_acf(new_data)
pacf_values=pacf(new_data,nlags=10)
plot_pacf(new_data)

In [ ]:
from statsmodels.tsa.arima.model import ARIMA

In [ ]:
train=new_df.iloc[:120]['Close']

In [ ]:
test=new_df.iloc[120:]

In [ ]:
model=ARIMA(train,order=(1,0,2))

In [ ]:
model_fit=model.fit()

In [ ]:
print(model_fit.summary())

In [ ]:
new_data['predicted']=model_fit.predict(start=len(train),end=len(train) + len(test)-1,dynamic=True)
new_data[['Close','predicted']].plot()

In [ ]:
from statsmodels.tsa.statespace.sarimax import SARIMAX

In [ ]:
model=SARIMAX(train,order=(1,0,2),seasonal_order=(1,0,2,12))
model_fit=model.fit()

In [ ]:
new_data['predicted']=model_fit.predict(start=len(train),end=len(train) + len(test)-1,dynamic=True)
new_data[['Close','predicted']].plot()

In [ ]:
forecast=model_fit.forecast(steps=60)
plt.figure(figsize=(10,6))
plt.plot(new_data.index,new_data['Close'],label='Original Data')

In [ ]:
plt.plot(forecast.index,forecast,label='Forecast',linestyle='--')
plt.xlabel("Date")
plt.ylabel("Close")
plt.title("Forecasting Stock Prices for the next five years")
plt.legend()